File: app/src/main/kotlin/tech/mmarca/openvitals/core/insights/PersonalBaseline.kt
package tech.mmarca.openvitals.core.insights

import java.time.LocalDate
import kotlin.math.abs
import kotlin.math.pow
import kotlin.math.sqrt

private const val MinimumBaselineSamples = 3
private const val AnomalyZScoreThreshold = 2.0
private const val BaselineTolerance = 0.0001
val PersonalBaselineWindows = listOf(30, 60, 90)

data class BaselineValue(
    val date: LocalDate,
    val value: Double,
)

data class BaselineSummary(
    val windowDays: Int,
    val average: Double,
    val standardDeviation: Double,
    val sampleCount: Int,
) {
    val usualLow: Double get() = average - standardDeviation
    val usualHigh: Double get() = average + standardDeviation
}

enum class BaselineStatus {
    USUAL,
    ABOVE,
    BELOW,
    UNUSUAL_HIGH,
    UNUSUAL_LOW,
}

data class PersonalBaselineInsight(
    val currentValue: Double,
    val primarySummary: BaselineSummary,
    val summaries: List<BaselineSummary>,
    val status: BaselineStatus,
) {
    val deviation: Double get() = currentValue - primarySummary.average
    val absoluteDeviation: Double get() = abs(deviation)
    val percentDeviation: Double?
        get() = primarySummary.average.takeIf { abs(it) > BaselineTolerance }
            ?.let { deviation / it * 100.0 }
    val isAnomaly: Boolean
        get() = status == BaselineStatus.UNUSUAL_HIGH || status == BaselineStatus.UNUSUAL_LOW
}

fun personalBaselineInsight(
    currentValue: Double,
    values: List<BaselineValue>,
    referenceDate: LocalDate,
    windows: List<Int> = PersonalBaselineWindows,
): PersonalBaselineInsight? {
    val summaries = windows
        .sorted()
        .mapNotNull { windowDays ->
            baselineSummary(
                windowDays = windowDays,
                values = values,
                referenceDate = referenceDate,
            )
        }

    val primarySummary = summaries.lastOrNull() ?: return null
    val status = baselineStatus(currentValue, primarySummary)
    return PersonalBaselineInsight(
        currentValue = currentValue,
        primarySummary = primarySummary,
        summaries = summaries,
        status = status,
    )
}

private fun baselineSummary(
    windowDays: Int,
    values: List<BaselineValue>,
    referenceDate: LocalDate,
): BaselineSummary? {
    val start = referenceDate.minusDays(windowDays.toLong() - 1)
    val windowValues = values
        .asSequence()
        .filter { it.date in start..referenceDate }
        .map { it.value }
        .filter { it > 0.0 }
        .toList()

    if (windowValues.size < MinimumBaselineSamples) return null

    val average = windowValues.average()
    val standardDeviation = standardDeviation(windowValues, average)
    return BaselineSummary(
        windowDays = windowDays,
        average = average,
        standardDeviation = standardDeviation,
        sampleCount = windowValues.size,
    )
}

private fun baselineStatus(
    currentValue: Double,
    summary: BaselineSummary,
): BaselineStatus {
    val standardDeviation = summary.standardDeviation
    if (standardDeviation <= BaselineTolerance) {
        return when {
            currentValue > summary.average + BaselineTolerance -> BaselineStatus.ABOVE
            currentValue < summary.average - BaselineTolerance -> BaselineStatus.BELOW
            else -> BaselineStatus.USUAL
        }
    }

    val zScore = (currentValue - summary.average) / standardDeviation
    return when {
        zScore >= AnomalyZScoreThreshold -> BaselineStatus.UNUSUAL_HIGH
        zScore <= -AnomalyZScoreThreshold -> BaselineStatus.UNUSUAL_LOW
        currentValue > summary.usualHigh -> BaselineStatus.ABOVE
        currentValue < summary.usualLow -> BaselineStatus.BELOW
        else -> BaselineStatus.USUAL
    }
}

private fun standardDeviation(values: List<Double>, average: Double): Double {
    val variance = values.sumOf { (it - average).pow(2) } / values.size
    return sqrt(variance)
}

File: lib/domain/insights/personal_baseline.dart
import 'dart:math' as math;

import 'package:freezed_annotation/freezed_annotation.dart';

import '../../core/time/local_date.dart';

part 'personal_baseline.freezed.dart';

const int _minimumBaselineSamples = 3;
const double _anomalyZScoreThreshold = 2.0;
const double _baselineTolerance = 0.0001;
const List<int> personalBaselineWindows = [30, 60, 90];

@freezed
abstract class BaselineValue with _$BaselineValue {
  const factory BaselineValue({
    required LocalDate date,
    required double value,
  }) = _BaselineValue;
}

@freezed
abstract class BaselineSummary with _$BaselineSummary {
  const BaselineSummary._();

  const factory BaselineSummary({
    required int windowDays,
    required double average,
    required double standardDeviation,
    required int sampleCount,
  }) = _BaselineSummary;

  double get usualLow => average - standardDeviation;
  double get usualHigh => average + standardDeviation;
}

enum BaselineStatus {
  usual,
  above,
  below,
  unusualHigh,
  unusualLow,
}

@freezed
abstract class PersonalBaselineInsight with _$PersonalBaselineInsight {
  const PersonalBaselineInsight._();

  const factory PersonalBaselineInsight({
    required double currentValue,
    required BaselineSummary primarySummary,
    required List<BaselineSummary> summaries,
    required BaselineStatus status,
  }) = _PersonalBaselineInsight;

  double get deviation => currentValue - primarySummary.average;
  double get absoluteDeviation => deviation.abs();
  double? get percentDeviation =>
      primarySummary.average.abs() > _baselineTolerance
          ? deviation / primarySummary.average * 100.0
          : null;
}

PersonalBaselineInsight? personalBaselineInsight(
  double currentValue,
  List<BaselineValue> values,
  LocalDate referenceDate, {
  List<int> windows = personalBaselineWindows,
}) {
  final sortedWindows = [...windows]..sort();
  final summaries = <BaselineSummary>[];
  for (final windowDays in sortedWindows) {
    final summary = _baselineSummary(windowDays, values, referenceDate);
    if (summary != null) summaries.add(summary);
  }

  if (summaries.isEmpty) return null;
  final primarySummary = summaries.last;
  final status = _baselineStatus(currentValue, primarySummary);
  return PersonalBaselineInsight(
    currentValue: currentValue,
    primarySummary: primarySummary,
    summaries: summaries,
    status: status,
  );
}

BaselineSummary? _baselineSummary(
  int windowDays,
  List<BaselineValue> values,
  LocalDate referenceDate,
) {
  final start = referenceDate.minusDays(windowDays - 1);
  final windowValues = values
      .where((value) => value.date.isBetween(start, referenceDate))
      .map((value) => value.value)
      .where((value) => value > 0.0)
      .toList();

  if (windowValues.length < _minimumBaselineSamples) return null;

  final average = _average(windowValues);
  final standardDeviation = _standardDeviation(windowValues, average);
  return BaselineSummary(
    windowDays: windowDays,
    average: average,
    standardDeviation: standardDeviation,
    sampleCount: windowValues.length,
  );
}

BaselineStatus _baselineStatus(double currentValue, BaselineSummary summary) {
  final standardDeviation = summary.standardDeviation;
  if (standardDeviation <= _baselineTolerance) {
    if (currentValue > summary.average + _baselineTolerance) {
      return BaselineStatus.above;
    }
    if (currentValue < summary.average - _baselineTolerance) {
      return BaselineStatus.below;
    }
    return BaselineStatus.usual;
  }

  final zScore = (currentValue - summary.average) / standardDeviation;
  if (zScore >= _anomalyZScoreThreshold) return BaselineStatus.unusualHigh;
  if (zScore <= -_anomalyZScoreThreshold) return BaselineStatus.unusualLow;
  if (currentValue > summary.usualHigh) return BaselineStatus.above;
  if (currentValue < summary.usualLow) return BaselineStatus.below;
  return BaselineStatus.usual;
}

double _average(List<double> values) =>
    values.fold<double>(0.0, (sum, value) => sum + value) / values.length;

double _standardDeviation(List<double> values, double average) {
  final variance = values.fold<double>(
        0.0,
        (sum, value) => sum + (value - average) * (value - average),
      ) /
      values.length;
  return math.sqrt(variance);
}
